### PR TITLE
FEATURE: Track API and user API requests

### PR DESCRIPTION
**Pull request by nbianca**
There are statistics for regular page views already, but none for API and user API requests.

### CODE REVIEW

**Comment by SamSaffron** [comment]
This is looking good, but I feel like a full integration test makes sense. 

1. make a request to `/about.json` or something with an good api key. Confirm it is logged. 
2. make a request to `/about.json` or something with a bad api key. Confirm it is not logged. 

Similar to (1) / (2) but using user api.

Generally I feel the correct implementation only logs proper API use so we need to be careful to test the permutations.

File: spec/integration/request_tracker_spec.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    )
+  end
+
+  let(:user_api_key) do
+    Fabricate(:user_api_key, scopes: [Fabricate.build(:user_api_key_scope, name: 'session_info')])
+  end
+
+  before do
+    ApplicationRequest.enable
+    CachedCounting.reset
```
Comment (tgxworld) [inline spec/integration/request_tracker_spec.rb:18]: Small one but it is good to reset this in the teardown as well.

File: spec/integration/request_tracker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    ApplicationRequest.disable
+    CachedCounting.disable
+  end
+
+  context 'when using an api key' do
+    it 'is counted as an API request' do
+      get "/u/#{api_key.user.username}.json", headers: { HTTP_API_KEY: api_key.key }
+      CachedCounting.flush
+
+      expect(response.status).to eq(200)
```
Comment (tgxworld) [inline spec/integration/request_tracker_spec.rb:32]:
```suggestion
      get "/u/#{api_key.user.username}.json", headers: { HTTP_API_KEY: api_key.key }
      expect(response.status).to eq(200)
      
      CachedCounting.flush
```

File: spec/integration/request_tracker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      expect(ApplicationRequest.api.first.count).to eq(1)
+    end
+  end
+
+  context 'when using an user api key' do
+    it 'is counted as a user API request' do
+      get '/session/current.json', headers: { HTTP_USER_API_KEY: user_api_key.key }
+      CachedCounting.flush
+
+      expect(response.status).to eq(200)
```
Comment (tgxworld) [inline spec/integration/request_tracker_spec.rb:46]:
```suggestion
      get '/session/current.json', headers: { HTTP_USER_API_KEY: user_api_key.key }
      expect(response.status).to eq(200)

      CachedCounting.flush
```

File: spec/lib/middleware/request_tracker_spec.rb [hunk 4]
```diff
@@ -100,6 +100,33 @@ def log_tracked_view(val)
       expect(ApplicationRequest.page_view_crawler.first.count).to eq(1)
     end
 
+    it "can log API requests correctly" do
```
Comment (tgxworld) [inline spec/lib/middleware/request_tracker_spec.rb:103]:
```suggestion
    it "logs API requests correctly" do
```

File: spec/models/report_spec.rb [hunk 5]
```diff
@@ -1233,6 +1233,53 @@ def expect_ignored_users_report_data_to_be_equal(data, user, ignores, mutes)
     end
   end
 
+  describe "consolidated_api_requests" do
```
Comment (tgxworld) [inline spec/models/report_spec.rb:1236]:
```suggestion
  describe ".report_consolidated_api_requests" do
```

**Review by tgxworld — APPROVED** [review]
I left some suggestions with regards to the tests but nothing blocking here.

**Pull request merged**